SIGSEGV when first refresh operation fails

## Bug Description

CloudSQL Proxy throws SIGSEGV once the first MySQL client connection is made and the proxy is started with "-ip_address_types=PRIVATE" and the CloudSQL instance does not have a private IP.  

```
./cloud_sql_proxy -instances=PROJECT_ID:us-central1:XYZ=tcp:3306 -verbose -ip_address_types=PRIVATE

2021/04/20 12:18:11 Rlimits for file descriptors set to {Current = 8500, Max = 9223372036854775807}
2021/04/20 12:18:11 using credential file for authentication; path="/Users/……./cred.json"
2021/04/20 12:18:12 Listening on 127.0.0.1:3306 for PROJECT_ID:us-central1:XYZ
2021/04/20 12:18:12 Ready for new connections
2021/04/20 12:18:22 New connection for "PROJECT_ID:us-central1:XYZ"
2021/04/20 12:18:22 refreshing ephemeral certificate for instance PROJECT_ID:us-central1:XYZ
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x155a56f]

goroutine 83 [running]:
github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy.(*Client).startRefresh.func1(0xc00032a540, 0xc0002e6180, 0x7ffeefbff5ad, 0x48, 0x45d964b800)
	/workspace/proxy/proxy/client.go:285 +0x38f
created by github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy.(*Client).startRefresh
	/workspace/proxy/proxy/client.go:264 +0x85


(venv) [user@machine:~/software] $ cloud_sql_proxy -version
Cloud SQL Auth proxy: 1.21.0+darwin.amd64
```

## How to reproduce (see above)

  1. Download and run proxy version 1.21.0+darwin.amd64
  2. Try to connect to DB without private IP via proxy (using private IP flag), resulting in signal SIGSEGV: segmentation violation

## Environment

1. OS type and version: <YOUR_OS_HERE>
2. Cloud SQL Proxy version (`./cloud_sql_proxy -version`): Cloud SQL Auth proxy: 1.21.0+darwin.amd64

--- a/proxy/proxy/client.go
+++ b/proxy/proxy/client.go
@@ -272,7 +272,7 @@ func (c *Client) startRefresh(instance string, refreshCfgBuffer time.Duration) c
 			logging.Errorf("failed to refresh the ephemeral certificate for %s, returning previous cert instead: %v", instance, err)
 			addr, cfg, ver, err = old.addr, old.cfg, old.version, old.err
 		}
-		c.cfgCache[instance] = cacheEntry{
+		e := cacheEntry{
 			lastRefreshed: time.Now(),
 			err:           err,
 			addr:          addr,
@@ -280,7 +280,15 @@ func (c *Client) startRefresh(instance string, refreshCfgBuffer time.Duration) c
 			cfg:           cfg,
 			done:          done,
 		}
+		c.cfgCache[instance] = e
 		c.cacheL.Unlock()
+
+		if !isValid(e) {
+			// Note: Future refreshes will not be scheduled unless another
+			// connection attempt is made.
+			logging.Errorf("failed to refresh the ephemeral certificate for %v: %v", instance, err)
+			return
+		}
 
 		certExpiration := cfg.Certificates[0].Leaf.NotAfter
 		now := time.Now()

--- a/proxy/proxy/client_test.go
+++ b/proxy/proxy/client_test.go
@@ -31,8 +31,8 @@ import (
 const instance = "instance-name"
 
 var (
-	errFakeDial = errors.New("this error is returned by the dialer")
-	forever     = time.Date(9999, 0, 0, 0, 0, 0, 0, time.UTC)
+	sentinelError = errors.New("sentinel error")
+	forever       = time.Date(9999, 0, 0, 0, 0, 0, 0, time.UTC)
 )
 
 type fakeCerts struct {
@@ -66,44 +66,46 @@ func (cs *blockingCertSource) Remote(instance string) (cert *x509.Certificate, a
 	return &x509.Certificate{}, "fake address", "fake name", "fake version", nil
 }
 
-func TestContextDialer(t *testing.T) {
-	b := &fakeCerts{}
-	c := &Client{
-		Certs: &blockingCertSource{
-			map[string]*fakeCerts{
-				instance: b,
-			},
-			forever,
+func newCertSource(certs *fakeCerts, expiration time.Time) CertSource {
+	return &blockingCertSource{
+		values: map[string]*fakeCerts{
+			instance: certs,
 		},
-		ContextDialer: func(context.Context, string, string) (net.Conn, error) {
-			return nil, errFakeDial
-		},
+		validUntil: expiration,
+	}
+}
+
+func newClient(cs CertSource) *Client {
+	return &Client{
+		Certs: cs,
 		Dialer: func(string, string) (net.Conn, error) {
-			return nil, fmt.Errorf("this dialer should not be used when ContextDialer is set")
+			return nil, sentinelError
 		},
 	}
+}
 
-	if _, err := c.DialContext(context.Background(), instance); err != errFakeDial {
+func TestContextDialer(t *testing.T) {
+	cs := newCertSource(&fakeCerts{}, forever)
+	c := newClient(cs)
+
+	c.ContextDialer = func(context.Context, string, string) (net.Conn, error) {
+		return nil, sentinelError
+	}
+	c.Dialer = func(string, string) (net.Conn, error) {
+		return nil, fmt.Errorf("this dialer should not be used when ContextDialer is set")
+	}
+
+	if _, err := c.DialContext(context.Background(), instance); err != sentinelError {
 		t.Errorf("unexpected error: %v", err)
 	}
 }
 
 func TestClientCache(t *testing.T) {
 	b := &fakeCerts{}
-	c := &Client{
-		Certs: &blockingCertSource{
-			map[string]*fakeCerts{
-				instance: b,
-			},
-			forever,
-		},
-		Dialer: func(string, string) (net.Conn, error) {
-			return nil, errFakeDial
-		},
-	}
+	c := newClient(newCertSource(b, forever))
 
 	for i := 0; i < 5; i++ {
-		if _, err := c.Dial(instance); err != errFakeDial {
+		if _, err := c.Dial(instance); err != sentinelError {
 			t.Errorf("unexpected error: %v", err)
 		}
 	}
@@ -117,17 +119,7 @@ func TestClientCache(t *testing.T) {
 
 func TestConcurrentRefresh(t *testing.T) {
 	b := &fakeCerts{}
-	c := &Client{
-		Certs: &blockingCertSource{
-			map[string]*fakeCerts{
-				instance: b,
-			},
-			forever,
-		},
-		Dialer: func(string, string) (net.Conn, error) {
-			return nil, errFakeDial
-		},
-	}
+	c := newClient(newCertSource(b, forever))
 
 	ch := make(chan error)
 	b.Lock()
@@ -144,7 +136,7 @@ func TestConcurrentRefresh(t *testing.T) {
 	b.Unlock()
 
 	for i := 0; i < numDials; i++ {
-		if err := <-ch; err != errFakeDial {
+		if err := <-ch; err != sentinelError {
 			t.Errorf("unexpected error: %v", err)
 		}
 	}
@@ -156,35 +148,31 @@ func TestConcurrentRefresh(t *testing.T) {
 }
 
 func TestMaximumConnectionsCount(t *testing.T) {
-	const maxConnections = 10
-	const numConnections = maxConnections + 1
-	var dials uint64 = 0
-
-	b := &fakeCerts{}
-	certSource := blockingCertSource{
-		map[string]*fakeCerts{},
-		forever,
+	certSource := &blockingCertSource{
+		values:     map[string]*fakeCerts{},
+		validUntil: forever,
 	}
+	c := newClient(certSource)
+
+	const maxConnections = 10
+	c.MaxConnections = maxConnections
+	var dials uint64
 	firstDialExited := make(chan struct{})
-	c := &Client{
-		Certs: &certSource,
-		Dialer: func(string, string) (net.Conn, error) {
-			atomic.AddUint64(&dials, 1)
-
-			// Wait until the first dial fails to ensure the max connections count is reached by a concurrent dialer
-			<-firstDialExited
-
-			return nil, errFakeDial
-		},
-		MaxConnections: maxConnections,
+	c.Dialer = func(string, string) (net.Conn, error) {
+		atomic.AddUint64(&dials, 1)
+		// Wait until the first dial fails to ensure the max connections count
+		// is reached by a concurrent dialer
+		<-firstDialExited
+		return nil, sentinelError
 	}
 
 	// Build certSource.values before creating goroutines to avoid concurrent map read and map write
+	const numConnections = maxConnections + 1
 	instanceNames := make([]string, numConnections)
 	for i := 0; i < numConnections; i++ {
 		// Vary instance name to bypass config cache and avoid second call to Client.tryConnect() in Client.Dial()
 		instanceName := fmt.Sprintf("%s-%d", instance, i)
-		certSource.values[instanceName] = b
+		certSource.values[instanceName] = &fakeCerts{}
 		instanceNames[i] = instanceName
 	}
 
@@ -218,18 +206,13 @@ func TestMaximumConnectionsCount(t *testing.T) {
 }
 
 func TestShutdownTerminatesEarly(t *testing.T) {
-	b := &fakeCerts{}
-	c := &Client{
-		Certs: &blockingCertSource{
-			map[string]*fakeCerts{
-				instance: b,
-			},
-			forever,
-		},
-		Dialer: func(string, string) (net.Conn, error) {
-			return nil, nil
-		},
+	cs := newCertSource(&fakeCerts{}, forever)
+	c := newClient(cs)
+	// Ensure the dialer returns no error.
+	c.Dialer = func(string, string) (net.Conn, error) {
+		return nil, nil
 	}
+
 	shutdown := make(chan bool, 1)
 	go func() {
 		c.Shutdown(1)
@@ -247,24 +230,16 @@ func TestShutdownTerminatesEarly(t *testing.T) {
 }
 
 func TestRefreshTimer(t *testing.T) {
-	timeToExpire := 5 * time.Second
-	b := &fakeCerts{}
+	timeToExpire := 2 * time.Second
 	certCreated := time.Now()
-	c := &Client{
-		Certs: &blockingCertSource{
-			map[string]*fakeCerts{
-				instance: b,
-			},
-			certCreated.Add(timeToExpire),
-		},
-		Dialer: func(string, string) (net.Conn, error) {
-			return nil, errFakeDial
-		},
-		RefreshCfgThrottle: 20 * time.Millisecond,
-		RefreshCfgBuffer:   time.Second,
-	}
+	cs := newCertSource(&fakeCerts{}, certCreated.Add(timeToExpire))
+	c := newClient(cs)
+
+	c.RefreshCfgThrottle = 20 * time.Millisecond
+	c.RefreshCfgBuffer = time.Second
+
 	// Call Dial to cache the cert.
-	if _, err := c.Dial(instance); err != errFakeDial {
+	if _, err := c.Dial(instance); err != sentinelError {
 		t.Fatalf("Dial(%s) failed: %v", instance, err)
 	}
 	c.cacheL.Lock()
@@ -288,9 +263,30 @@ func TestRefreshTimer(t *testing.T) {
 }
 
 func TestSyncAtomicAlignment(t *testing.T) {
-	// The sync/atomic pkg has a bug that requires the developer to guarantee 64-bit alignment when using 64-bit functions on 32-bit systems.
+	// The sync/atomic pkg has a bug that requires the developer to guarantee
+	// 64-bit alignment when using 64-bit functions on 32-bit systems.
 	c := &Client{}
 	if a := unsafe.Offsetof(c.ConnectionsCounter); a%64 != 0 {
 		t.Errorf("Client.ConnectionsCounter is not aligned: want %v, got %v", 0, a)
 	}
+}
+
+type invalidRemoteCertSource struct{}
+
+func (cs *invalidRemoteCertSource) Local(instance string) (tls.Certificate, error) {
+	return tls.Certificate{}, nil
+}
+
+func (cs *invalidRemoteCertSource) Remote(instance string) (*x509.Certificate, string, string, string, error) {
+	return nil, "", "", "", sentinelError
+}
+
+func TestRemoteCertError(t *testing.T) {
+	c := newClient(&invalidRemoteCertSource{})
+
+	_, err := c.DialContext(context.Background(), instance)
+	if err != sentinelError {
+		t.Errorf("expected sentinel error, got %v", err)
+	}
+
 }

